public/app/features/alerting/unified/hooks/useFilteredRules.ts

```diff
@@ -84,3 +82,21 @@ const reduceGroups = (filters: RuleFilterState) => {
     return groupAcc;
   };
 };
+
+const isQueryingDataSource = (rulerRule: RulerGrafanaRuleDTO, filter: RuleFilterState): boolean => {
+  if (!filter.dataSource) {
+    return true;
+  }
+
+  if (!Array.isArray(rulerRule.grafana_alert.data)) {
```

Comment:
Imho it's not worthwhile to pepper code guarding against `null` that is not normally expected and does not match types. It's boilerplaty, have to do it in a lot of places, hides the problem rather than fixes it.

I suggest to either:
* Just let it fail to become aware of the problem faster, and fix the backend issue
* fill in the possible null at the api endpoint level, in `api/ruler.ts`